### PR TITLE
Service improvements

### PR DESCRIPTION
**Pull request by permutohedron**
BeaconService is now both started (with START_STICKY flag) and bound when used for background scanning (through the RegionBootstrap). It prevents scanning from being stopped when user swipes out the task or the system kills the service for resources. BeaconService is going to be immediately restarted. Also when background the state of beacon monitoring is preserved upon service restarts. (So that the double didEnterRegion() triggering on Service recreation is not an issue any more).

### CODE REVIEW

**Comment by davidgyoung** [comment]
Thanks for the PR, @permutohedron.  Some interesting changes here, so I need to take some time to review.

File: src/main/java/org/altbeacon/beacon/service/BeaconService.java [hunk 1]
```diff
@@ -237,17 +257,37 @@ public void onDestroy() {
         LogManager.i(TAG, "onDestroy called.  stopping scanning");
         handler.removeCallbacksAndMessages(null);
         mCycledScanner.stop();
+        monitoringStatus.stopStatusPreservationOnProcessDestruction();
+        monitoringStatus.clear();
+    }
+
+    @Override
+    public void onTaskRemoved(Intent rootIntent) {
```
Comment (permutohedron) [inline src/main/java/org/altbeacon/beacon/service/BeaconService.java]: Called when service is started. (When it is only bound that is not invoked)

**Comment by davidgyoung** [comment]
Excellent pull request, @	permutohedron.  A few small items need to be cleaned up, but then I'm inclined to merge this.  This warrants a new dot release because preserving state of region monitoring across service restarts is a significant new feature.

Because this is a significant change,  I'd probably release this as a 2.8-beta first, to see if other folks report problems with it before releasing it to everyone on jCenter.

Can you please take a crack the small changes mentioned in my comments, so I can merge?

Apologies for it taking me so long to review this.

File: src/main/java/org/altbeacon/beacon/service/BeaconService.java [hunk 2]
```diff
@@ -487,20 +488,13 @@ protected void onProgressUpdate(Void... values) {
...
-                // Need to check if region is null in case it was removed from the collection by
-                // another thread during iteration
-                if (region != null) {
-                    if (region.matchesBeacon(beacon)) {
-                        matched.add(region);
-                    } else {
-                        LogManager.d(TAG, "This region (%s) does not match beacon: %s", region, beacon);
-                    }
-                }
+        for (Region region : regions) {
```
Comment (permutohedron) [inline src/main/java/org/altbeacon/beacon/service/BeaconService.java]: Removed unnecessary null checking of region.

File: src/main/java/org/altbeacon/beacon/service/BeaconService.java [hunk 3]
```diff
@@ -237,17 +256,36 @@ public void onDestroy() {
         LogManager.i(TAG, "onDestroy called.  stopping scanning");
         handler.removeCallbacksAndMessages(null);
         mCycledScanner.stop();
+        monitoringStatus.stopStatusPreservationOnProcessDestruction();
```
Comment (permutohedron) [inline src/main/java/org/altbeacon/beacon/service/BeaconService.java]: Removed unnecessary monitoringStatus.clear()

File: src/main/java/org/altbeacon/beacon/service/MonitoringStatus.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
+package org.altbeacon.beacon.service;
```
Comment (permutohedron) [inline src/main/java/org/altbeacon/beacon/service/MonitoringStatus.java]: Moved to proper package

File: src/main/java/org/altbeacon/beacon/service/MonitoringStatus.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+public class MonitoringStatus {
+    private static MonitoringStatus sInstance;
+    private static final String TAG = MonitoringStatus.class.getSimpleName();
+    public static final String STATUS_PRESERVATION_FILE_NAME =
+            "org.altbeacon.beacon.service.monitoring_status_state";
+    private final Map<Region, RegionMonitoringState> mRegionsStatesMap
+            = new HashMap<Region, RegionMonitoringState>();
+
+    private Context mContext;
+
```
Comment (permutohedron) [inline src/main/java/org/altbeacon/beacon/service/MonitoringStatus.java]: Android field naming convention done ;)

File: src/main/java/org/altbeacon/beacon/service/MonitoringStatus.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            } else {
+                LogManager.d(TAG, "This region (%s) does not match beacon: %s", region, beacon);
+            }
+        }
+        return matched;
+    }
+
+    private void saveMonitoringStatusIfOn() {
+        if(!mStatePreservationIsOn) return;
+        LogManager.d(TAG, "saveMonitoringStatusIfOn()");
```
Comment (permutohedron) [inline src/main/java/org/altbeacon/beacon/service/MonitoringStatus.java]: Now it's debug not error log.

**Comment by permutohedron** [comment]
Hello @davidgyoung Great to hear that pull request is good :) Made some changes you've mentioned.
I also must apologize for it took me so long to do it. I've been realy busy these days ;/
Hope you are still planning merge this to the repo :)

**Pull request merged**